File: compiler/test/dotty/tools/vulpix/ChildJVMMain.java
package dotty.tools.vulpix;

import java.io.File;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.lang.reflect.Method;

public class ChildJVMMain {
    static final String MessageStart = "##THIS IS THE START FOR ME, HELLO##";
    static final String MessageEnd = "##THIS IS THE END FOR ME, GOODBYE##";

    private static void runMain(String dir) throws Exception {
        Method meth = null;
        Object[] args = new Object[]{ new String[]{ } };
        try {
            String jcp = System.getProperty("java.class.path");
            String sep = File.pathSeparator;
            System.setProperty("java.class.path", jcp == null ? dir : dir + sep + jcp);

            ArrayList<URL> cp = new ArrayList<>();
            for (String path : dir.split(sep))
                cp.add(new File(path).toURI().toURL());

            URLClassLoader ucl = new URLClassLoader(cp.toArray(new URL[cp.size()]));

            Class<?> cls = ucl.loadClass("Test");
            meth = cls.getMethod("main", String[].class);
        }
        catch (Throwable e) {
            // Include the failure stack trace to the test output
            System.out.println(MessageStart);
            e.printStackTrace();
            throw e;
        }
        System.out.println(MessageStart);

        meth.invoke(null, args);
    }

    public static void main(String[] args) throws Exception {
      BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));

      while (true) {
          runMain(stdin.readLine());
          System.out.println(MessageEnd);
      }
    }
}

File: compiler/test/dotty/tools/vulpix/ChildJVMMain.scala
package dotty.tools.vulpix

import java.io.{BufferedReader, File, InputStreamReader}
import java.net.{URL, URLClassLoader}
import java.util.ArrayList

object ChildJVMMain:
  val MessageStart = "##THIS IS THE START FOR ME, HELLO##"
  val MessageEnd = "##THIS IS THE END FOR ME, GOODBYE##"

  def runMain(dir: String): Unit =
    def meth =
      val jcp = System.getProperty("java.class.path")
      val sep = File.pathSeparator
      System.setProperty("java.class.path", if jcp == null then dir else dir + sep + jcp)

      val loader =
        val cp = ArrayList[URL]()
        val paths = dir.split(sep)
        for path <- paths do
          cp.add(File(path).toURI().toURL())
        val urls = cp.toArray(Array.ofDim[URL](cp.size))
        URLClassLoader(urls)

      val cls = loader.loadClass("Test")
      cls.getMethod("main", classOf[Array[String]])
    end meth
    val m =
      try meth
      catch t =>
        // Include the failure stack trace to the test output
        System.out.println(MessageStart)
        t.printStackTrace()
        throw t
    System.out.println(MessageStart)
    m.invoke(null, Array.empty[String])

  def main(args: Array[String]): Unit =
    inline def savingSystem[T](inline body: => T): T =
      val savedIn  = System.in
      val savedOut = System.out
      val savedErr = System.err
      try body
      finally
        System.setIn(savedIn)
        System.setOut(savedOut)
        System.setErr(savedErr)
    val stdin = BufferedReader(InputStreamReader(System.in))
    while true do
      savingSystem:
        runMain(stdin.readLine())
      println(MessageEnd)
